packages/ai-jsx/src/core/conversation.tsx

```diff
@@ -238,6 +238,45 @@ function toConversationMessages(partialRendering: AI.PartiallyRendered[]): Conve
   });
 }
 
+async function loggableMessage(
+  message: ConversationMessage,
+  render: AI.RenderContext['render'],
+  measureCost?: (message: ConversationMessage, render: AI.ComponentContext['render']) => Promise<number>
+) {
+  let textPromise: PromiseLike<string> | undefined = undefined;
+  switch (message.type) {
+    case 'user':
+    case 'assistant':
+    case 'system':
+      textPromise = render(message.element);
```

Comment:
what's the cost of rendering in this function?  will it start a chain of other renders if part of the tree has not rendered? Or is the nature of when this is called gaurantee that the element already has been rendered and is now memoized?

Reply:
Good question! It's certainly possible for this to block on rendering that hasn't yet happened. But in practice, we already do that when logging conversations to measure token cost, and this will only happen if the caller requests the conversation to be logged (i.e. the pre-model/post-model invocations).

So in practice, it won't cause any rendering to happen meaningfully earlier than it already would.